File: 2-sorting/2-quick-sort/main.js
function quickSort(arr, l, r) {
  if (l == undefined) {
    l = 0;
  }

  if (r == undefined) {
    r = arr.length - 1;
  }

  if (l >= r) {
    return;
  }

  let pivot = Math.floor((Math.random() * (r - l + 1))) + l;
  let x = arr[r];
  [arr[pivot], arr[r]] = [arr[r], arr[pivot]];

  let j = l;

  for (let i = l; i <= r; ++i) {
    if (arr[i] < x) {
      [arr[i], arr[j]] = [arr[j], arr[i]];
      j++;
    }
  }

  [arr[r], arr[j]] = [arr[j], arr[r]];
  quickSort(arr, l, j - 1);
  quickSort(arr, j + 1, r);
}

let a = [1, 8, 3, 2, 2, 7, 9, 5];
quickSort(a);

console.log(a);

File: 2-sorting/2-quick-sort/main.cpp
#include <climits>
#include <iostream>
#include <vector>

using namespace std;

void quickSort(vector<int>& v, int l, int r) {
  if (l >= r) {
    return;
  }

  int pivot = (rand() % (r - l + 1)) + l;
  int x = v[pivot];

  swap(v[pivot], v[r]);

  int j = l;
  for (int i = l; i < r; ++i) {
    if (v[i] < x) {
      swap(v[i], v[j++]);
    }
  }

  swap(v[j], v[r]);
  quickSort(v, l, j - 1);
  quickSort(v, j + 1, r);
}

void quickSort(vector<int>& v) { quickSort(v, 0, v.size() - 1); }

int main() {
  vector<int> v = {1, 8, 3, 2, 2, 7, 9, 5};

  quickSort(v);

  for (int n : v) {
    cout << n << " ";
  }
  cout << endl;
  return 0;
}
